patterns/multisig/proxy.ts

```diff
@@ -0,0 +1,136 @@
+import { types } from "polkadot_dev/mod.ts"
+import { Proxy, Utility } from "polkadot_dev/mod.ts"
+import { Event } from "polkadot_dev/types/pallet_proxy/pallet.ts"
+import { RuntimeEvent } from "polkadot_dev/types/polkadot_runtime.ts"
+import { ArrayRune, MultiAddress, Rune, RunicArgs } from "../../mod.ts"
+import { hex, Sr25519 } from "../../util/mod.ts"
+
+export interface BuildPureProxyMultisigProps {
+  sender: Sr25519
+  admins: Uint8Array[]
+  threshold?: number
+}
+
+export function buildPureProxyMultisig<X1>(props: RunicArgs<X1, BuildPureProxyMultisigProps>) {
+  // TODO: clean up with https://github.com/paritytech/capi/issues/589
+  const { admins, threshold: thresholdOrUndef, sender } = RunicArgs.resolve(props)
+  const adminCount = admins.map((admins) => admins.length)
+
+  // TODO: validate threshold within acceptable range
+  const threshold = Rune
+    .resolve(thresholdOrUndef)
+    .unhandle(undefined)
+    .rehandle(undefined, () => adminCount)
+
+  // Mapping from signatory public key to to-be-created stash account index
+  const adminDisambiguationIndicesMap = admins.map((a) =>
+    Object.fromEntries(a.map((publicKey, i) => [hex.encode(publicKey), i]))
+  )
+
+  const createPureCalls = adminCount
+    .map((n) => Array.from({ length: n + 1 }, (_, i) => i))
+    .into(ArrayRune)
+    .mapArray((index) =>
+      Proxy.createPure({
+        proxyType: "Any",
+        delay: 0,
+        index,
+      })
+    )
+
+  const proxyAccountIds = Utility
+    .batch({ calls: createPureCalls })
+    .signed({ sender })
+    .sent()
+    .dbgStatus("Creating pure proxies:")
+    .txEvents()
+    .pipe(filterPureCreatedEvents)
+    .map((e): Record<number, Uint8Array> =>
+      Object.fromEntries(e.map(({ disambiguationIndex, pure }) => [disambiguationIndex, pure]))
+    )
+
+  const stashMultiAddress = Rune
+    .tuple([proxyAccountIds, adminCount])
+    .map(([rec, i]) => rec[i]!)
+    .map(MultiAddress.Id)
+
+  const proxySignatoryAccountIds = Rune
+    .tuple([adminCount, proxyAccountIds])
+    .map(([stashIndex, indices]) => {
+      const { [stashIndex]: _, ...rest } = indices
+      return Object.values(rest)
+    })
+    .into(ArrayRune)
+
+  const adminProxies = Rune.tuple([adminDisambiguationIndicesMap, proxyAccountIds])
+    .map(([adminDisambiguationIndicesMap, proxyAddressMap]): Record<string, Uint8Array> =>
+      Object.fromEntries(
+        Object
+          .keys(adminDisambiguationIndicesMap)
+          .map((sig) => [sig, proxyAddressMap[adminDisambiguationIndicesMap[sig]!]!]),
+      )
+    )
+
+  return Rune.rec({
+    multisig: Rune.rec({
+      signatories: proxySignatoryAccountIds,
+      threshold,
+    }),
+    stashMultiAddress,
+    adminProxies,
+  })
+}
+
+export function addProxy<X>(
```

Comment:
Can we move these proxy-only utilities into a `patterns/proxy/mod.ts`?

Reply:
Might be worth creating a simple proxy-only example as well

Reply:
> Might be worth creating a simple proxy-only example as well

Can do in a separate issue/pr

Reply:
https://github.com/paritytech/capi/issues/645